THRIFT-3891 TNonblockingServer configured with more than one IO threads does not always return from serve() upon stop()

jeking3 opened the pull request:
I took the comments from @nsuke and applied them to the PR.  I was able to reasonably eliminate the boolean you mentioned however I could not go back to loopbreak.  I see no reasonable way to synchronize here.  The run() thread will eventually need to release a mutex and THEN call the event loop.  There's no way inside breakLoop() to know if it is actually IN the event loop.  Given this, the exit variant is the only safe method to break the loop.

The obsoletes PR #1080.

jeking3 commented:
Looks like I ran the build and tests outside the dockerfile so it didn't do any libevent stuff. :|

nsuke commented on lib/cpp/src/thrift/server/TNonblockingServer.cpp:
```diff
@@ -212,7 +212,8 @@ class TNonblockingServer::TConnection {
               const sockaddr* addr,
-              socklen_t addrLen) {
+              socklen_t addrLen)
+  {
```
Enter key by accident ?

nsuke commented on lib/cpp/src/thrift/server/TNonblockingServer.cpp:
```diff
@@ -1520,13 +1524,9 @@ void TNonblockingIOThread::breakLoop(bool error) {
-  // sets a flag so that the loop exits on the next event
-  event_base_loopbreak(eventBase_);
+  // cause the loop to stop ASAP - even if it has things to do in it
+  event_base_loopexit(eventBase_, 0);
```
+1
According to another look at the header doc, loopbreak is typically to be called from inside the loop.

nsuke commented on lib/cpp/src/thrift/server/TNonblockingServer.cpp:
```diff
@@ -820,9 +821,11 @@ void TNonblockingServer::TConnection::setFlags(short eventFlags) {
-  if (event_del(&event_) == -1) {
-    GlobalOutput.perror("TConnection::close() event_del", THRIFT_GET_SOCKET_ERROR);
+  if (eventFlags_) {
+    if (event_del(&event_) == -1) {
```
nits: single `if` with `eventFlags_ && event_del...` ?

nsuke commented on lib/cpp/src/thrift/server/TNonblockingServer.cpp:
```diff
@@ -820,9 +821,11 @@ void TNonblockingServer::TConnection::setFlags(short eventFlags) {
-    GlobalOutput.perror("TConnection::close() event_del", THRIFT_GET_SOCKET_ERROR);
+  if (eventFlags_) {
+    if (event_del(&event_) == -1) {
+      GlobalOutput("TConnection::close event_del");
```
Not sure why changing from `GlobalOutput.perror` to `GlobalOutput`.

jeking3 replied:
Sorry I copied this from another location and it must have been using GlobalOutput.  I will fix this.

jeking3 commented:
The failure looks like an inability to stop the loop.
I an going to look into changing the nonblocking server's stop() method to inject an event that causes it to call loopbreak from within the loop.

jeking3 commented:
In the next push I have changed the way the nonblocking server test starts up, avoiding a ~100ms delay three times.  Then I changed how the loop stops so that it always stops from inside using loopbreak.

jeking3 commented:
@nsuke to test this I ran TNonblockingServerTest in a tight loop for 30 minutes.  It never made it past 30 seconds before, and that was with 300ms delay built into each test.  The test runs much faster now because I properly synchronized the "wait for the server to be listening" part of the test, and I fixed the shutdown mechanism so that it always occurs from within the event loop.

jeking3 commented:
@nsuke this is ready for review again.

bgedik commented on lib/cpp/src/thrift/server/TNonblockingServer.cpp:
```diff
@@ -1566,24 +1562,26 @@ void TNonblockingIOThread::setCurrentThreadHighPriority(bool value) {
+    cleanupEvents();
+  }
 
   GlobalOutput.printf("TNonblockingServer: IO thread #%d run() done!", number_);
```
Is there any way to control the outputting of these messages? When we run with TNonblockingServer, we get these output. I couldn't find a way to mask them.

jeking3 replied:
You can install your own GlobalOutput handler.  The hooks should be in the top level thrift header.

bgedik replied:
Thanks, but for a message like {{"TNonblockingServer: IO thread #0 run() done!"}} is there any kind of metadata that tells me if this is a info message or a warning/error message. It seems to be being output without any information attached to it. That makes it hard to filter the messages that we consider unnecessary.

jeking3 replied:
All of these messages were already in the implementation and they do not appear to have any discrimination as to the severity.  My assumption is they are all meant for "printf style debugging".

jeking3 commented:
Once someone has a chance to approve the changes I can merge.

jeking3 commented:
Could someone please do a code review and approve or request changes?  Thanks.
